<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

## Lab: Cleaning Rock Song Data

_Authors: Dave Yerrington (SF)_

---


In [50]:
import pandas as pd
import numpy as np 
import seaborn as sns

%matplotlib inline

### 1. Load `rock.csv` and do an initial examination of its data columns.

In [51]:
rockfile = "../datasets/rock.csv"

In [52]:
# Load the data.

df = pd.read_csv(rockfile)

In [53]:
# Look at the information regarding its columns.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2230 entries, 0 to 2229
Data columns (total 8 columns):
Song Clean      2230 non-null object
ARTIST CLEAN    2230 non-null object
Release Year    1653 non-null object
COMBINED        2230 non-null object
First?          2230 non-null int64
Year?           2230 non-null int64
PlayCount       2230 non-null int64
F*G             2230 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.4+ KB


In [54]:
df.dtypes

Song Clean      object
ARTIST CLEAN    object
Release Year    object
COMBINED        object
First?           int64
Year?            int64
PlayCount        int64
F*G              int64
dtype: object

In [55]:
df.sample(10)

Song Clean    ARTIST CLEAN Release Year  \
2210  Beer Drinkers & Hell Raisers          ZZ Top         1973   
1729           Don't Bring Me Down     The Animals         1966   
154                     Rebel Yell      Billy Idol         1984   
1467                 Gimme Shelter  Rolling Stones         1969   
1826          The Fool On The Hill     The Beatles         1967   
1699           Life During Wartime   Talking Heads         1979   
1592               Pulling Mussels         Squeeze         1980   
2148                 Summer Nights       Van Halen         1986   
1550                   The Warrior         Scandal         1984   
31                        Sin City           AC/DC          NaN   

                                    COMBINED  First?  Year?  PlayCount  F*G  
2210  Beer Drinkers & Hell Raisers by ZZ Top       1      1          8    8  
1729      Don't Bring Me Down by The Animals       1      1          1    1  
154                 Rebel Yell by Billy Idol       1      1         89   89  
1467         Gimme Shelter by Rolling Stones       1      1         96   96  
1826     The Fool On The Hill by The Beatles       1      1          2    2  
1699    Life During Wartime by Talking Heads       1      1         23   23  
1592              Pulling Mussels by Squeeze       1      1          2    2  
2148              Summer Nights by Van Halen       1      1          4    4  
1550                  The Warrior by Scandal       1      1          5    5  
31                         Sin City by AC/DC       1      0          1    0

### 2.  Clean up the column names.

Let's clean up the column names. There are two ways we can accomplish this:

#### 2.A Change the column names when you import the data using `pd.read_csv()`.

Notice that, when passing `names=[..A LIST OF STRING..]` with a number of columns that matches the number of strings in the passed list, you replace the column names.

NOTE: When you create custom column names, the first row of the `.csv` already represents a header. It is important to tell `pandas` to skip that row. The `skiprows=1` keyword argument to `read_csv()` will tell `pandas` to skip the first row.

In [56]:
df.columns

Index([u'Song Clean', u'ARTIST CLEAN', u'Release Year', u'COMBINED', u'First?',
       u'Year?', u'PlayCount', u'F*G'],
      dtype='object')

In [57]:
import string
# the string library has default strings that contain all letters or numbers
uppercase = string.ascii_uppercase
lowercase = string.ascii_lowercase

df.columns = [''.join([c.lower() for c in column if c in uppercase+lowercase]) for column in df.columns]

In [58]:
df.columns

Index([u'songclean', u'artistclean', u'releaseyear', u'combined', u'first',
       u'year', u'playcount', u'fg'],
      dtype='object')

#### 2.B Change column names using the `.rename()` function.

The `.rename()` function takes an argument, `columns=name_dict`, in which `name_dict` is a dictionary containing the original column names as keys and the new column names as values.

In [59]:
# Change the column names using the `.rename()` function.
df = df.rename(columns = lambda column : ''.join([c.lower() for c in column if c in uppercase+lowercase]))

In [60]:
df.head(2)

songclean  artistclean releaseyear                         combined  \
0  Caught Up in You  .38 Special        1982  Caught Up in You by .38 Special   
1      Fantasy Girl  .38 Special         NaN      Fantasy Girl by .38 Special   

   first  year  playcount  fg  
0      1     1         82  82  
1      1     0          3   0

#### 2.C Reassigning the `.columns` attribute of a DataFrame.

You can also just reassign the `.columns` attribute to a list of strings containing the new column names. 

The only caveat with reassigning `.columns` is that you have to reassign all of the column names at once. You can't partially replace a value by working on `.columns` directly. You have to reassign `.columns` with a list of equal length. 

In [61]:
# Replace the column names by reassigning the `.columns` attribute.
df.columns = [''.join([c.lower() for c in column if c in uppercase+lowercase]) for column in df.columns]

In [62]:
df.columns

Index([u'songclean', u'artistclean', u'releaseyear', u'combined', u'first',
       u'year', u'playcount', u'fg'],
      dtype='object')

### 3. Subsetting data where null values exist.

We have mixed `str` and `NaN` values in the `release` column. `NaN` stands for "not a number" and is the way `pandas` handles "nulls" or nonexistent data. We can use the `.isnull()` method of a Series to find null values.

Print the header of the data subset to where the `release` column is null values.

In [63]:
df.columns

Index([u'songclean', u'artistclean', u'releaseyear', u'combined', u'first',
       u'year', u'playcount', u'fg'],
      dtype='object')

In [64]:
# Show records where df['release'] is null
df[df['releaseyear'].isnull()].head()

songclean  artistclean releaseyear  \
1             Fantasy Girl  .38 Special         NaN   
10   Baby, Please Don't Go        AC/DC         NaN   
13  CAN'T STOP ROCK'N'ROLL        AC/DC         NaN   
16        Girls Got Rhythm        AC/DC         NaN   
24         Let's Get It Up        AC/DC         NaN   

                           combined  first  year  playcount  fg  
1       Fantasy Girl by .38 Special      1     0          3   0  
10   Baby, Please Don't Go by AC/DC      1     0          1   0  
13  CAN'T STOP ROCK'N'ROLL by AC/DC      1     0          5   0  
16        Girls Got Rhythm by AC/DC      1     0         24   0  
24         Let's Get It Up by AC/DC      1     0          4   0

In [65]:
df[~df['releaseyear'].isnull()].head()

songclean   artistclean releaseyear  \
0        Caught Up in You   .38 Special        1982   
2         Hold On Loosely   .38 Special        1981   
3  Rockin' Into the Night   .38 Special        1980   
4       Art For Arts Sake          10cc        1975   
5              Kryptonite  3 Doors Down        2000   

                                combined  first  year  playcount  fg  
0        Caught Up in You by .38 Special      1     1         82  82  
2         Hold On Loosely by .38 Special      1     1         85  85  
3  Rockin' Into the Night by .38 Special      1     1         18  18  
4              Art For Arts Sake by 10cc      1     1          1   1  
5             Kryptonite by 3 Doors Down      1     1         13  13

### 4. Update slices of your DataFrame based on mask selection/slices.

In many scenarios, we want to upate values in our DataFrame according to criteria. Let's say we wanted to set all of the null values in `release` to 0.

With newer versions of `pandas`, in order to manipulate data in the original DataFrame, we have to use `.loc` while performing reassignment using a mask and an index.

For example, the following won't always work:
```python
df[row_mask]['column_name'] = new_value
```

The best way to accomplish the same task is:
```python
df.loc[row_mask, 'column_name'] = new_value
```

For multiple column assignment, you would use:
```python
df.loc[row_mask, ['col_1', 'col_2', 'col_3']] = new_value
```

#### 4.A Let's try it out. Make all of the null values in `release` 0.

In [66]:
# Replace release nulls with 0

In [67]:
df[df['releaseyear'].isnull()].index

Int64Index([   1,   10,   13,   16,   24,   25,   26,   29,   31,   35,
            ...
            2203, 2205, 2206, 2208, 2215, 2216, 2218, 2221, 2222, 2225],
           dtype='int64', length=577)

In [68]:
df.loc[df[df['releaseyear'].isnull()].index, 'releaseyear'] = 0

In [69]:
# cheat df['releaseyear'] = df['releaseyear'].fillna(0)

#### 4.B Verify that `release` contains no null values.

In [70]:
df[df['releaseyear'].isnull()]

Empty DataFrame
Columns: [songclean, artistclean, releaseyear, combined, first, year, playcount, fg]
Index: []

### 5. Ensure that the data types of the columns make sense. 

Verifying column data types is a critical part of data munging. If columns have the wrong data type, then there is usually corrupted or incorrect data in some of the observations.

#### 5.A Look at the data types for the columns. Are any incorrect given what the data represents?

In [71]:
# A:

df.dtypes

songclean      object
artistclean    object
releaseyear    object
combined       object
first           int64
year            int64
playcount       int64
fg              int64
dtype: object

In [72]:
df.sample(5)

songclean             artistclean releaseyear  \
809           What You Need                    INXS        1985   
1218  Mama, I'm Coming Home           Ozzy Osbourne        1991   
1267     With a Little Luck  Paul McCartney & Wings           0   
751               Civil War           Guns N' Roses        1991   
799         ALL YOU ZOMBIES                 Hooters        1985   

                                          combined  first  year  playcount  fg  
809                          What You Need by INXS      1     1         28  28  
1218        Mama, I'm Coming Home by Ozzy Osbourne      1     1         31  31  
1267  With a Little Luck by Paul McCartney & Wings      1     0          2   0  
751                     Civil War by Guns N' Roses      1     1          1   1  
799                     ALL YOU ZOMBIES by Hooters      1     1          4   4

##### COMMENT: releaseyear as object?!?!

### 6. Investigate and clean up the `release` column.

The `release` column is a string data type when it should be an integer.

#### 6.A Figure out what value(s) are causing the `release` column to be encoded as a string instead of an integer.

In [73]:
df['releaseyear'].unique()

array(['1982', 0, '1981', '1980', '1975', '2000', '2002', '1992', '1985',
       '1993', '1976', '1995', '1979', '1984', '1977', '1990', '1986',
       '1974', '2014', '1987', '1973', '2001', '1989', '1997', '1971',
       '1972', '1994', '1970', '1966', '1965', '1983', '1955', '1978',
       '1969', '1999', '1968', '1988', '1962', '2007', '1967', '1958',
       '1071', '1996', '1991', '2005', '2011', '2004', '2012', '2003',
       '1998', '2008', '1964', '2013', '2006', 'SONGFACTS.COM', '1963',
       '1961'], dtype=object)

In [74]:
for c in df['releaseyear']:
    try:
        float(c)
    except:
        print(c)

SONGFACTS.COM


#### 6.B Look at the rows in which there is incorrect data in the `release` column.

In [75]:
df[df['releaseyear']=='SONGFACTS.COM']

songclean     artistclean    releaseyear  \
1504  Bullfrog Blues  Rory Gallagher  SONGFACTS.COM   

                              combined  first  year  playcount  fg  
1504  Bullfrog Blues by Rory Gallagher      1     1          1   1

#### 6.C. Clean up the data. Normally we may replace the offending data with null np.nan values, however we previously converted all of the nan values in the release column to zeros so we might as well continue with the same practice. Replacing with 0 (or nan) will allow us to convert the column to numeric.

In [76]:
df.loc[df[df['releaseyear']=='SONGFACTS.COM'].index, 'releaseyear'] = np.nan

In [77]:
df['releaseyear'] = df['releaseyear'].astype(float)

In [78]:
df.dtypes

songclean       object
artistclean     object
releaseyear    float64
combined        object
first            int64
year             int64
playcount        int64
fg               int64
dtype: object

### 7. Get summary statistics for the `release` column using the `.describe()` function.

Now that the `release` column is finally a numeric data type, we can apply the `.describe()` function.  

#### 7.A Print out the summary stats for the `release` column. What is the earliest and latest release date?

In [79]:
# A:
df['releaseyear'].describe()

count    2229.000000
mean     1465.988784
std       866.834789
min         0.000000
25%         0.000000
50%      1973.000000
75%      1981.000000
max      2014.000000
Name: releaseyear, dtype: float64

In [80]:
df[df['releaseyear']>0]['releaseyear'].describe()

count    1652.000000
mean     1978.019976
std        24.191247
min      1071.000000
25%      1971.000000
50%      1977.000000
75%      1984.000000
max      2014.000000
Name: releaseyear, dtype: float64

#### 7.B Based on the summary statistics, is there anything else wrong with the `release` column? 

_Looking at the DataFrame that contains the year 1071, we can see that the year was probably corrupted and should be replaced with something else if possible._

### 8. Make changes and investigate using custom functions with `.apply()`.

Let's say we want to traverse every single row in our data set and apply a function to that row.

#### 8.A Write a function that will take a row of a DataFrame and print out the song, artist, and whether or not the release date is < 1970.


In [81]:
# A:

def print_info(row):
    print(row['songclean'],row['artistclean'],row['releaseyear']<1970)

#### 8.B Using the `.apply()` function, apply the function you wrote to the first four rows of the DataFrame.

You will need to tell the `apply` function to operate row by row. Setting the keyword argument as `axis=1` indicates that the function should be applied to each row individually.

In [82]:
# THIS IS BY COLUMNS --> ERROR!
# df.apply(print_info, axis=0)

In [83]:
# This is by rows
df.apply(print_info, axis=1)

('Caught Up in You', '.38 Special', False)
('Fantasy Girl', '.38 Special', True)
('Hold On Loosely', '.38 Special', False)
("Rockin' Into the Night", '.38 Special', False)
('Art For Arts Sake', '10cc', False)
('Kryptonite', '3 Doors Down', False)
('Loser', '3 Doors Down', False)
("When I'm Gone", '3 Doors Down', False)
("What's Up?", '4 Non Blondes', False)
('Take On Me', 'a-ha', False)
("Baby, Please Don't Go", 'AC/DC', True)
('Back In Black', 'AC/DC', False)
('Big Gun', 'AC/DC', False)
("CAN'T STOP ROCK'N'ROLL", 'AC/DC', True)
('Dirty Deeds Done Dirt Cheap', 'AC/DC', False)
('For Those About To Rock', 'AC/DC', False)
('Girls Got Rhythm', 'AC/DC', True)
('Hard As A Rock', 'AC/DC', False)
('Have a Drink On Me', 'AC/DC', False)
('Hells Bells', 'AC/DC', False)
('Highway To Hell', 'AC/DC', False)
("It's A Long Way To The Top", 'AC/DC', False)
('Jailbreak', 'AC/DC', False)
('Let There Be Rock', 'AC/DC', False)
("Let's Get It Up", 'AC/DC', True)
('Live Wire', 'AC/DC', True)
('Moneytalks', '

('Small Town', 'John Mellencamp', False)
('What If I Came Knocking', 'John Mellencamp', True)
('Illegal Smile', 'John Prine', True)
('I CAN SEE CLEARLY NOW', 'Johnny Nash', False)
('Brickyard Road', 'Johnny Van Zant', False)
('Shanty', 'Jonathan Edwards', True)
('Roadrunner', 'Jonathan Richman', False)
('Any Way You Want It', 'Journey', False)
('City of the Angels', 'Journey', True)
("Don't Stop Believin'", 'Journey', False)
('Escape', 'Journey', True)
('Faithfully', 'Journey', False)
('Feeling That Way', 'Journey', False)
('Just the Same Way', 'Journey', True)
('Lights', 'Journey', False)
('Look Into The Future', 'Journey', True)
("Lovin', Touchin' Squeezin'", 'Journey', False)
('Only the Young', 'Journey', False)
('Open Arms', 'Journey', False)
('Send Her My Love', 'Journey', False)
('Separate Ways (Worlds Apart)', 'Journey', False)
('Stone In Love', 'Journey', True)
('Wheel in the Sky', 'Journey', False)
("Who's Crying Now", 'Journey', False)
('Breaking The Law', 'Judas Priest', Fal

('Oh, Well', 'Rockets', True)
("Tired Of Toein' The Line", 'Rocky Burnette', True)
('Every Picture Tells A Story', 'Rod Stewart', False)
('Hot Legs', 'Rod Stewart', True)
("I Know I'm Losing You", 'Rod Stewart', True)
('Maggie May', 'Rod Stewart', False)
('Reason to Believe', 'Rod Stewart', False)
('Stay with Me', 'Rod Stewart', True)
('The First Cut Is The Deepest', 'Rod Stewart', False)
("Tonight's the Night (Gonna Be Alright)", 'Rod Stewart', False)
('You Wear It Well', 'Rod Stewart', False)
("You're In My Heart (The Final Acclaim)", 'Rod Stewart', False)
('Free Me', 'Roger Daltry', True)
('Had a Dream (Sleeping With the Enemy)', 'Roger Hodgson', True)
('The Logical Song', 'Roger Hodgson', False)
("(I Can't Get No) Satisfaction", 'Rolling Stones', True)
('19th Nervous Breakdown', 'Rolling Stones', True)
("Ain't Too Proud to Beg", 'Rolling Stones', False)
('Almost Hear You Sigh', 'Rolling Stones', False)
('Angie', 'Rolling Stones', False)
('Beast Of Burden', 'Rolling Stones', False)


('Lola', 'The Kinks', False)
('Sunny Afternoon', 'The Kinks', True)
('Superman', 'The Kinks', True)
('Tired of Waiting', 'The Kinks', True)
('Victoria', 'The Kinks', True)
('Waterloo Sunset', 'The Kinks', True)
('You Really Got Me', 'The Kinks', True)
('My Sharona', 'The Knack', False)
('Mrs. Robinson', 'The Lemonheads', True)
('Summer In the City', "The Lovin' Spoonful", True)
("CALIFORNIA DREAMIN'", 'The Mamas ', True)
("Can't You See", 'The Marshall Tucker Band', False)
('HEARD IT IN A LOVE SONG', 'The Marshall Tucker Band', False)
('Take The Highway', 'The Marshall Tucker Band', True)
('Gypsy', 'The Moody Blues', True)
('Question', 'The Moody Blues', False)
('Ride My See Saw', 'The Moody Blues', True)
('The Story In Your Eyes', 'The Moody Blues', False)
('Your Wildest Dreams', 'The Moody Blues', False)
('Life in the Fast Lane', 'The New Eagles', False)
('Come Out and Play', 'The Offspring', False)
('Self Esteem', 'The Offspring', False)
("The Kids Aren't Alright", 'The Offspring', 

0       None
1       None
2       None
3       None
4       None
5       None
6       None
7       None
8       None
9       None
10      None
11      None
12      None
13      None
14      None
15      None
16      None
17      None
18      None
19      None
20      None
21      None
22      None
23      None
24      None
25      None
26      None
27      None
28      None
29      None
        ... 
2200    None
2201    None
2202    None
2203    None
2204    None
2205    None
2206    None
2207    None
2208    None
2209    None
2210    None
2211    None
2212    None
2213    None
2214    None
2215    None
2216    None
2217    None
2218    None
2219    None
2220    None
2221    None
2222    None
2223    None
2224    None
2225    None
2226    None
2227    None
2228    None
2229    None
Length: 2230, dtype: object

You'll notice that there will be a final output Series of `None` values. The `.apply()` function, if a return value is not specified, will return a Series of `None` values (similar to how the default return for Python functions is `None` when a return statement is not specified).

### 9. Write a function that converts cells in a DataFrame to float and otherwise replaces them with `np.nan`.

If applied to our data, it would keep only the numeric information and otherwise input null values.

Recall that the try-except syntax in Python is a great way to try something and take another action if the initial step fails:

```python
try:
    Perform some action.
except:
   Perform some other action if the first failed with an error.
```

#### 9.A Write the function that takes a column and converts all of its values to float if possible and `np.nan` otherwise. The return value should be the converted Series.

In [84]:
# A:

def float_or_nan(column):
    new_col = []
    for el in column:
        try:
            new_col.append(float(el))
        except:
            new_col.append(np.nan)
    return new_col

In [85]:
def element_converter(element):
    try:
        return float(element)
    except:
        return np.nan

def float_or_nan(column):
    return column.map(element_converter)

#### 9.B Try your function out on the rock song data and ensure the output is what you expected.


In [86]:
# A:

df.apply(float_or_nan)

songclean  artistclean  releaseyear  combined  first  year  playcount  \
0           NaN          NaN       1982.0       NaN    1.0   1.0       82.0   
1           NaN          NaN          0.0       NaN    1.0   0.0        3.0   
2           NaN          NaN       1981.0       NaN    1.0   1.0       85.0   
3           NaN          NaN       1980.0       NaN    1.0   1.0       18.0   
4           NaN          NaN       1975.0       NaN    1.0   1.0        1.0   
5           NaN          NaN       2000.0       NaN    1.0   1.0       13.0   
6           NaN          NaN       2000.0       NaN    1.0   1.0        1.0   
7           NaN          NaN       2002.0       NaN    1.0   1.0        6.0   
8           NaN          NaN       1992.0       NaN    1.0   1.0        3.0   
9           NaN          NaN       1985.0       NaN    1.0   1.0        1.0   
10          NaN          NaN          0.0       NaN    1.0   0.0        1.0   
11          NaN          NaN       1980.0       NaN    1.0   1.0       97.0   
12          NaN          NaN       1993.0       NaN    1.0   1.0        6.0   
13          NaN          NaN          0.0       NaN    1.0   0.0        5.0   
14          NaN          NaN       1976.0       NaN    1.0   1.0       85.0   
15          NaN          NaN       1981.0       NaN    1.0   1.0       46.0   
16          NaN          NaN          0.0       NaN    1.0   0.0       24.0   
17          NaN          NaN       1995.0       NaN    1.0   1.0        1.0   
18          NaN          NaN       1980.0       NaN    1.0   1.0       39.0   
19          NaN          NaN       1980.0       NaN    1.0   1.0       74.0   
20          NaN          NaN       1979.0       NaN    1.0   1.0       92.0   
21          NaN          NaN       1975.0       NaN    1.0   1.0       39.0   
22          NaN          NaN       1984.0       NaN    1.0   1.0        1.0   
23          NaN          NaN       1977.0       NaN    1.0   1.0        3.0   
24          NaN          NaN          0.0       NaN    1.0   0.0        4.0   
25          NaN          NaN          0.0       NaN    1.0   0.0        2.0   
26          NaN          NaN          0.0       NaN    1.0   0.0       20.0   
27          NaN          NaN       1979.0       NaN    1.0   1.0        1.0   
28          NaN          NaN       1980.0       NaN    1.0   1.0       21.0   
29          NaN          NaN          0.0       NaN    1.0   0.0       45.0   
...         ...          ...          ...       ...    ...   ...        ...   
2200        NaN          NaN       1983.0       NaN    1.0   1.0       81.0   
2201        NaN          NaN       1971.0       NaN    1.0   1.0       44.0   
2202        NaN          NaN       1971.0       NaN    1.0   1.0        1.0   
2203        NaN          NaN          0.0       NaN    1.0   0.0        5.0   
2204        NaN          NaN       1971.0       NaN    1.0   1.0        1.0   
2205        NaN          NaN          0.0       NaN    1.0   0.0        2.0   
2206        NaN          NaN          0.0       NaN    1.0   0.0        1.0   
2207        NaN          NaN       1968.0       NaN    1.0   1.0       14.0   
2208        NaN          NaN          0.0       NaN    1.0   0.0        1.0   
2209        NaN          NaN       1976.0       NaN    1.0   1.0        1.0   
2210        NaN          NaN       1973.0       NaN    1.0   1.0        8.0   
2211        NaN          NaN       1980.0       NaN    1.0   1.0       34.0   
2212        NaN          NaN       1983.0       NaN    1.0   1.0       86.0   
2213        NaN          NaN       1983.0       NaN    1.0   1.0       29.0   
2214        NaN          NaN       1975.0       NaN    1.0   1.0        2.0   
2215        NaN          NaN          0.0       NaN    1.0   0.0       16.0   
2216        NaN          NaN          0.0       NaN    1.0   0.0       10.0   
2217        NaN          NaN       1973.0       NaN    1.0   1.0        6.0   
2218        NaN          NaN          0.0       NaN    1

#### 9.C Describe the new float-only DataFrame.

In [87]:
df.apply(float_or_nan).describe()

songclean  artistclean  releaseyear  combined   first         year  \
count     2.000000          0.0  2229.000000       0.0  2230.0  2230.000000   
mean   1012.000000          NaN  1465.988784       NaN     1.0     0.741256   
std    1367.544515          NaN   866.834789       NaN     0.0     0.438043   
min      45.000000          NaN     0.000000       NaN     1.0     0.000000   
25%     528.500000          NaN     0.000000       NaN     1.0     0.000000   
50%    1012.000000          NaN  1973.000000       NaN     1.0     1.000000   
75%    1495.500000          NaN  1981.000000       NaN     1.0     1.000000   
max    1979.000000          NaN  2014.000000       NaN     1.0     1.000000   

         playcount           fg  
count  2230.000000  2230.000000  
mean     16.872646    15.048430  
std      25.302972    25.288366  
min       0.000000     0.000000  
25%       1.000000     0.000000  
50%       4.000000     3.000000  
75%      21.000000    18.000000  
max     142.000000   142.000000